src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -1076,6 +1102,34 @@ private bool AddIfNotPresent(T value)
             return true;
         }
 
+        private void AddValue(int hashCode, T value)
+        {
+            int bucket = hashCode % _buckets.Length;
+
+#if DEBUG
+            for (int i = _buckets[bucket] - 1; i >= 0; i = _slots[i].next)
+            {
+                Debug.Assert(_slots[i].hashCode != hashCode || !_comparer.Equals(_slots[i].value, value));
+            }
+#endif
```

Comment:
Would it be worth adding a `Debug.Assert(hashCode == InternalGetHashCode(value));` as well?